README.md
---
Version: 1.2.1 
Title: M365 Apps Intune scripted dynamic install using Office Deployment Toolkit
Authors: JankeSkanke, sandytsang
Owner: JankeSkanke
Date: 16/12/2022
Description: Deploying M365 with Intune dynamically as a Win32App
---
  [![HitCount](https://hits.dwyl.com/msendpointmgr/m365apps.svg?style=flat)](http://hits.dwyl.com/msendpointmgr/m365apps)

# M365 Apps Intune scripted dynamic install using Office Deployment Toolkit 

## This solution covers installation of the following products

* [M365 Apps(Office)](#main-office-package-getting-configuration-from-external-url)
* [Project](#project-and-visio)
* [Visio](#project-and-visio)
* [Proofing tools](#proofing-tools)

Each product is made of the following components

* Install script (PowerShell)
* Local Configuration.xml or External URL
* Detection (script or documented)

***
## Main Office Package (getting configuration from External URL)

1. Define your config XML (Example below, can be generated at office.com)

```xml
<Configuration ID="9aa11e20-2e29-451a-b0ba-f1ae3e89d18d">
  <Add OfficeClientEdition="64" Channel="MonthlyEnterprise" MigrateArch="TRUE">
    <Product ID="O365ProPlusRetail">
      <Language ID="en-us" />
      <ExcludeApp ID="Groove" />
      <ExcludeApp ID="Lync" />
      <ExcludeApp ID="Bing" />
    </Product>
  </Add>
  <Property Name="SharedComputerLicensing" Value="0" />
  <Property Name="FORCEAPPSHUTDOWN" Value="TRUE" />
  <Property Name="DeviceBasedLicensing" Value="0" />
  <Property Name="SCLCacheOverride" Value="0" />
  <Updates Enabled="TRUE" />
  <AppSettings>
    <Setup Name="Company" Value="Company Name" />
  </AppSettings>
  <Display Level="None" AcceptEULA="FALSE" />
</Configuration>
```

1. Create a .Intunewim using the Win32 Content Prep tool [Prepare Win32 app content for upload](https://learn.microsoft.com/en-us/mem/intune/apps/apps-win32-prepare?WT.mc_id=EM-MVP-5002085) containing the InstallM365Apps.ps1 
2. Upload .Intunewim and define the following parameters during install 
    * Install Command:
      * ```powershell.exe -executionpolicy bypass -file InstallM365Apps.ps1 -XMLURL "https://mydomain.com/xmlfile.xml"```
    * Uninstall Command:
      * ```powershell.exe -executionpolicy bypass -file InstallM365Apps.ps1``` (Not working yet)

    <img src="/.images/officeinstall.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">

    * Install behaviour: System 
    * Requirements (probable 64 bit Windows something)
    * Detection: Use PowerShell detection Script M365AppsWin32DetectionScript.ps1 
  1. Assign

***

## Main Office Package (using configuration.xml inside package)

1. Define your config XML (Example below, can be generated at office.com)
```xml
<Configuration ID="9aa11e20-2e29-451a-b0ba-f1ae3e89d18d">
  <Add OfficeClientEdition="64" Channel="MonthlyEnterprise" MigrateArch="TRUE">
    <Product ID="O365ProPlusRetail">
      <Language ID="en-us" />
      <ExcludeApp ID="Groove" />
      <ExcludeApp ID="Lync" />
      <ExcludeApp ID="Bing" />
    </Product>
  </Add>
  <Property Name="SharedComputerLicensing" Value="0" />
  <Property Name="FORCEAPPSHUTDOWN" Value="TRUE" />
  <Property Name="DeviceBasedLicensing" Value="0" />
  <Property Name="SCLCacheOverride" Value="0" />
  <Updates Enabled="TRUE" />
  <AppSettings>
    <Setup Name="Company" Value="Company Name" />
  </AppSettings>
  <Display Level="None" AcceptEULA="FALSE" />
</Configuration>
```
2. Create a .Intunewim using the Win32 Content Prep tool [Prepare Win32 app content for upload](https://learn.microsoft.com/en-us/mem/intune/apps/apps-win32-prepare?WT.mc_id=EM-MVP-5002085) containing the configuration.xml and the InstallM365Apps.ps1 
2. Upload .Intunewim and define the following parameters during install 
    * Install Command:
      * ```powershell.exe -executionpolicy bypass -file InstallM365Apps.ps1```
    * Uninstall Command:
      * ```powershell.exe -executionpolicy bypass -file InstallM365Apps.ps1``` (Not working yet)
    * Install behaviour: System
    * Requirements (probable 64 bit Windows something)
    * Detection: Use PowerShell detection Script M365AppsWin32DetectionScript.ps1
  1. Assign

> Note: If you are deploying other languages than EN-US, you might need to edit the script M365AppsWin32DetectionScript.ps1 to properly detect your installation.

***
## Project and Visio

1. Define your config XML 
```xml
<Configuration ID="fc6a02c8-622f-4cf4-bf7f-6c57847b0580">
  <Add OfficeClientEdition="64" Version="MatchInstalled">
    <Product ID="ProjectProRetail">
      <Language ID="MatchInstalled" Fallback="en-us" TargetProduct="O365ProPlusRetail"/>
      <ExcludeApp ID="Groove" />
      <ExcludeApp ID="OneDrive" />
    </Product>
  </Add>
  <Property Name="FORCEAPPSHUTDOWN" Value="TRUE" />
  <Updates Enabled="TRUE" />
  <AppSettings>
    <Setup Name="Company" Value="Company Name" />
  </AppSettings>
  <Display Level="None" AcceptEULA="TRUE" />
</Configuration>
```
> The provided configuration.xml can not be imported on config.office.com and must be stored in for instance a Azure blob storage when using external url option. 

* This configuration file example will match the language installed (M365 Apps). 
* TargetProduct is required for that to work. 
* This example will shutdown running Office Apps for end users during install. 
* Follow others steps from the main office package. 
 <img src="/.images/visioinstall.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">

Project Install Command (Local):

```PowerShell
powershell.exe -executionpolicy bypass -file InstallProject.ps1 
```

Project Install Command (ExternalXML):

```PowerShell
powershell.exe -executionpolicy bypass -file InstallProject.ps1 -XMLURL "https://mydomain.com/xmlfile.xml"
```

Visio Install Command (Local):

```PowerShell
powershell.exe -executionpolicy bypass -file InstallVisio.ps1 
```

Visio Install Command (ExternalXML):

```PowerShell
powershell.exe -executionpolicy bypass -file InstallVisio.ps1 -XMLURL "https://mydomain.com/xmlfile.xml"
```

***

## Proofing tools or LanguagePacks

We recommend installing only 1 language on the computers unless your requirements are very specific. But there might still be some users that would need a complete language pack or proofing tools for various languages. The main thinking here is to have all possible proofing tools in your environment as available to end user to install by their own choosing.

For these 2 options the included configuration.xml files are just "templates" as the script it self will rewrite the XML dynamically based on the parameters you send to the script.

>There is no need to maintain this XML as long as Microsoft does not change the XML requirements.

***
**EXAMPLES for LanguagePacks**

```PowerShell
powershell.exe -executionpolicy bypass -file InstallLanguagePacks.ps1 -LanguageID nb-no -Action Install
powershell.exe -executionpolicy bypass -file InstallLanguagePacks.ps1 -LanguageID nb-no -Action Uninstall
```


**EXAMPLES for Proofing Tools**

```PowerShell
powershell.exe -executionpolicy bypass -file InstallProofingTools.ps1 -LanguageID nb-no -Action Install
powershell.exe -executionpolicy bypass -file InstallProofingTools.ps1 -LanguageID nb-no -Action Uninstall
```

<img src="/.images/proofinginstall_1.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">
<img src="/.images/proofinginstall_2.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">

***

It is also recommended that you have a requirement to check if Main Office is installed on the device as the install will fail if you try to install the a language pack or proofing tools without Office installed. This can be done using a registry key check or using the provided requirement script.

<img src="/.images/proofingrequire.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">

**ProofingRequirementScript.ps1**

***

Detection of language packs is best using registry detection. Example Norwegian: 

* Key Path: **HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\O365ProPlusRetail - nb-no**
* Value Name: **DisplayName**
* String Comparison - Equals
* Microsoft 365 Apps for enterprise - nb-no

Detection of the proofing tools can be done either with the provided detection script, customized for each LanguageID or by using a registry key check(Recommended).

EXAMPLE Detection Rule:
Registry
<img src="/.images/proofingdetect.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">

**HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\O365ProPlusRetail - nb-no.proof**

***
Get URL from config.office.com : 

<img src="/.images/configofficeurl.png" alt="Office Install XML" title="Office Install XML" style="display: inline-block; margin: 0 auto; max-width: 300px">


For more details and instructions go to [MSEndpointMgr Blog](https://msendpointmgr.com/2022/10/23/installing-m365-apps-as-win32-app-in-intune/)

This solution has been developed by @JankeSkanke with assistance from @sandytsang and @maurice-daly



CHANGELOG.MD
# Changelog

## 1.2.1 - 16.12.2022

 * Added verification of the setup.exe downloaded by verify signature on file and root trust to certificate
   * Verify signed by Microsoft
   * Verify Chain and check if Root is Microsoft
   * Verify root certificate exists in local Certificate Root Store

## 1.2 - 23.11.2022

* Changed to Evergreen Url for setup.exe to avoid any issues with "scraping" the ODT download site and simplification. (Requested by [@aaronparker](https://github.com/aaronparker))
* Added cleanup action to remove temporary setup files after installation
* Added Languagepacks as an separated option. Install with parameters same way as Proofing tools. The XML files in the folder is only templates used by script to configure given language.

Example

```PowerShell
powershell.exe -executionpolicy bypass -file InstallLanguagePacks.ps1 -LanguageID nb-no -Action Install
powershell.exe -executionpolicy bypass -file InstallLanguagePacks.ps1 -LanguageID nb-no -Action Uninstall
```

***

## 1.1 - 28.10.2022

* Added support for external XML configuration.xml for M365 Apps main package, Visio and Project as an option
* No change for Proofing Tools

Example

Without external XML (Requires configuration.xml in the package) - Same as Version 1.0

```PowerShell
powershell.exe -executionpolicy bypass -file <installfile>.ps1
```

With external XML (Requires XML to be provided by URL)  - New option in the scripts

```PowerShell
powershell.exe -executionpolicy bypass -file <installfile>.ps1 -XMLURL "https://mydomain.com/xmlfile.xml"
```

***

## 1.0 Initial release - 23.10.2022

No additional information - see Readme

***


ProofingTools/InstallProofingTools.ps1
<#
.SYNOPSIS
  Script to install Proofingtools as a Win32 App 

.DESCRIPTION
    Script to install Proofingtools as a Win32 App by downloading the latest Office Deployment Toolkit
    Running Setup.exe from downloaded files with provided config.xml file. 

.PARAMETER LanguageID
    Set the language ID in the correct formatting (like nb-no or en-us)
.PARAMETER Action 
    Supported actions are Install or Uninstall 

.EXAMPLE 
    InstallProofingTools.ps1 -LanguageID "nb-no" -Action Install
    InstallProofingTools.ps1 -LanguageID "nb-no" -Action Uninstall

.NOTES
    Version:          1.2
    Author:           Jan Ketil Skanke
    Creation Date:    01.07.2021
    Purpose/Change:   Initial script development
    Author:           Jan Ketil Skanke
    Contributor       Sandy Zeng 
    Contact:          @JankeSkanke @sandytsang
    Updated:     2022-22-09
    Version history:
        1.0 - (2020-10-11) Script created
        1.0.1 - (2022-15-06) MultiLanguageSupport via parameter 
        1.0.2 - (2022-20-06) Dynamically change configuration.xml based on LanguageID Parameter
        1.1 - (2022-22-09) Adding support for uninstall, now equires 2 xml files - uninstall.xml and install.xml
        1.2.0 - (2022-23-11) Moved from ODT download to Evergreen url for setup.exe
        1.2.1 - (2022-01-12) Adding function to validate signing on downloaded setup.exe
#>

#region parameters
[CmdletBinding()]
Param (
    [Parameter(Mandatory=$true)]
    [string]$LanguageID,

    [parameter(Mandatory=$true)]
    [ValidateNotNullOrEmpty()]
    [ValidateSet("Install", "Uninstall")]
    [string]$Action 
)
#endregion parameters

#Region Functions
function Write-LogEntry {
	param (
		[parameter(Mandatory = $true, HelpMessage = "Value added to the log file.")]
		[ValidateNotNullOrEmpty()]
		[string]$Value,
		[parameter(Mandatory = $true, HelpMessage = "Severity for the log entry. 1 for Informational, 2 for Warning and 3 for Error.")]
		[ValidateNotNullOrEmpty()]
		[ValidateSet("1", "2", "3")]
		[string]$Severity,
		[parameter(Mandatory = $false, HelpMessage = "Name of the log file that the entry will written to.")]
		[ValidateNotNullOrEmpty()]
		[string]$FileName = $LogFileName
	)
	# Determine log file location
	$LogFilePath = Join-Path -Path $env:SystemRoot -ChildPath $("Temp\$FileName")
	
	# Construct time stamp for log entry
	$Time = -join @((Get-Date -Format "HH:mm:ss.fff"), " ", (Get-WmiObject -Class Win32_TimeZone | Select-Object -ExpandProperty Bias))
	
	# Construct date for log entry
	$Date = (Get-Date -Format "MM-dd-yyyy")
	
	# Construct context for log entry
	$Context = $([System.Security.Principal.WindowsIdentity]::GetCurrent().Name)
	
	# Construct final log entry
	$LogText = "<![LOG[$($Value)]LOG]!><time=""$($Time)"" date=""$($Date)"" component=""$($LogFileName)"" context=""$($Context)"" type=""$($Severity)"" thread=""$($PID)"" file="""">"
	
	# Add value to log file
	try {
		Out-File -InputObject $LogText -Append -NoClobber -Encoding Default -FilePath $LogFilePath -ErrorAction Stop
		if ($Severity -eq 1) {
			Write-Verbose -Message $Value
		} elseif ($Severity -eq 3) {
			Write-Warning -Message $Value
		}
	} catch [System.Exception] {
		Write-Warning -Message "Unable to append log entry to $LogFileName.log file. Error message at line $($_.InvocationInfo.ScriptLineNumber): $($_.Exception.Message)"
	}
}
function Start-DownloadFile {
    param(
        [parameter(Mandatory=$true)]
        [ValidateNotNullOrEmpty()]
        [string]$URL,

        [parameter(Mandatory=$true)]
        [ValidateNotNullOrEmpty()]
        [string]$Path,

        [parameter(Mandatory=$true)]
        [ValidateNotNullOrEmpty()]
        [string]$Name
    )
    Begin {
        # Construct WebClient object
        $WebClient = New-Object -TypeName System.Net.WebClient
    }
    Process {
        # Create path if it doesn't exist
        if (-not(Test-Path -Path $Path)) {
            New-Item -Path $Path -ItemType Directory -Force | Out-Null
        }

        # Start download of file
        $WebClient.DownloadFile($URL, (Join-Path -Path $Path -ChildPath $Name))
    }
    End {
        # Dispose of the WebClient object
        $WebClient.Dispose()
    }
}
function Invoke-XMLUpdate  {
    param(
        [parameter(Mandatory=$true)]
        [ValidateNotNullOrEmpty()]
        [string]$LanguageID,

        [parameter(Mandatory=$true)]
        [ValidateNotNullOrEmpty()]
        [string]$Filename,

        [parameter(Mandatory=$true)]
        [ValidateNotNullOrEmpty()]
		[ValidateSet("Install", "Uninstall")]
        [string]$Action 
    )
    if ($Action -eq "Install"){
        $xmlDoc = [System.Xml.XmlDocument](Get-Content $FileName)
        $xmlDoc.Configuration.Add.Product.Language.ID = $LanguageID
        $xmlDoc.Save($FileName); 
    }
    else {
        $xmlDoc = [System.Xml.XmlDocument](Get-Content $FileName)
        $xmlDoc.Configuration.Remove.Product.Language.ID = $LanguageID
        $xmlDoc.Save($FileName);
    }
}
function Invoke-FileCertVerification {
    param(
        [parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [string]$FilePath
    )
    # Get a X590Certificate2 certificate object for a file
    $Cert = (Get-AuthenticodeSignature -FilePath $FilePath).SignerCertificate
    $CertStatus = (Get-AuthenticodeSignature -FilePath $FilePath).Status
    if ($Cert){
        #Verify signed by Microsoft and Validity
        if ($cert.Subject -match "O=Microsoft Corporation" -and $CertStatus -eq "Valid"){
            #Verify Chain and check if Root is Microsoft
            $chain = New-Object -TypeName System.Security.Cryptography.X509Certificates.X509Chain
            $chain.Build($cert) | Out-Null
            $RootCert = $chain.ChainElements | ForEach-Object {$_.Certificate}| Where-Object {$PSItem.Subject -match "CN=Microsoft Root"}
            if (-not [string ]::IsNullOrEmpty($RootCert)){
                #Verify root certificate exists in local Root Store
                $TrustedRoot = Get-ChildItem -Path "Cert:\LocalMachine\Root" -Recurse | Where-Object { $PSItem.Thumbprint -eq $RootCert.Thumbprint}
                if (-not [string]::IsNullOrEmpty($TrustedRoot)){
                    Write-LogEntry -Value "Verified setupfile signed by : $($Cert.Issuer)" -Severity 1
                    Return $True
                }
                else {
                    Write-LogEntry -Value  "No trust found to root cert - aborting" -Severity 2
                    Return $False
                }
            }
            else {
                Write-LogEntry -Value "Certificate chain not verified to Microsoft - aborting" -Severity 2 
                Return $False
            }
        }
        else {
            Write-LogEntry -Value "Certificate not valid or not signed by Microsoft - aborting" -Severity 2 
            Return $False
        }  
    }
    else {
        Write-LogEntry -Value "Setup file not signed - aborting" -Severity 2
        Return $False
    }
}
#Endregion Functions

#Region Initialisations
$LogFileName = "M365AppsSetup.log"
#Endregion Initialisations
switch -Wildcard ($Action) { 
    {($PSItem -match "Install")}{
      $FileName = "install.xml"
    }
    {($PSItem -match "Uninstall")}{
        $FileName = "uninstall.xml"
    }
}

#Initate Install
Write-LogEntry -Value "Initiating Proofing tools $($LanguageID) $($Action) process" -Severity 1
#Attempt Cleanup of SetupFolder
if (Test-Path "$($env:SystemRoot)\Temp\OfficeSetup"){
    Remove-Item -Path "$($env:SystemRoot)\Temp\OfficeSetup" -Recurse -Force -ErrorAction SilentlyContinue
}

$SetupFolder = (New-Item -ItemType "directory" -Path "$($env:SystemRoot)\Temp" -Name OfficeSetup -Force).FullName

try{
    #Download latest Office setup.exe
    $SetupEverGreenURL = "https://officecdn.microsoft.com/pr/wsus/setup.exe"
    Write-LogEntry -Value "Attempting to download latest Office setup executable" -Severity 1
    Start-DownloadFile -URL $SetupEverGreenURL -Path $SetupFolder -Name "setup.exe"
    
    try{
         #Start install preparations
         $SetupFilePath = Join-Path -Path $SetupFolder -ChildPath "setup.exe"
         if (-Not (Test-Path $SetupFilePath)) {
             Throw "Error: Setup file not found"
         }
         Write-LogEntry -Value "Setup file found at $($SetupFilePath)" -Severity 1
        try{
            #Prepare Proofing tools installation or removal
            $OfficeCR2Version = [System.Diagnostics.FileVersionInfo]::GetVersionInfo("$($SetupFolder)\setup.exe").FileVersion 
            Write-LogEntry -Value "Office C2R Setup is running version $OfficeCR2Version" -Severity 1
            if (Invoke-FileCertVerification -FilePath $SetupFilePath){
                Invoke-XMLUpdate -LanguageID $LanguageID -Filename "$($PSScriptRoot)\$($Filename)" -Action $Action
                Copy-Item "$($PSScriptRoot)\$($Filename)" $SetupFolder -Force -ErrorAction Stop
                Write-LogEntry -Value "Proofing tools $($LanguageID) configuration file copied" -Severity 1           
                Try{
                    #Running office installer
                    Write-LogEntry -Value "Starting Proofing tools $($LanguageID) $($Action) with Win32App method" -Severity 1
                    $OfficeInstall = Start-Process $SetupFilePath -ArgumentList "/configure $($SetupFolder)\$($Filename)" -NoNewWindow -Wait -PassThru -ErrorAction Stop
                }
                catch [System.Exception]{
                    Write-LogEntry -Value  "Error running the Proofing tools $($LanguageID) $($Action). Errormessage: $($_.Exception.Message)" -Severity 3
                }
            }
            else {
                Throw "Error: Unable to verify setup file signature"
            }
        }
        catch [System.Exception]{
            Write-LogEntry -Value  "Error preparing Proofing tools $($LanguageID) $($Action). Errormessage: $($_.Exception.Message)" -Severity 3
        }
    }
    catch [System.Exception]{
        Write-LogEntry -Value  "Error finding office setup file. Errormessage: $($_.Exception.Message)" -Severity 3
    }
    
}
catch [System.Exception]{
    Write-LogEntry -Value  "Error downloading Office setup file. Errormessage: $($_.Exception.Message)" -Severity 3
}
#Cleanup 
if (Test-Path "$($env:SystemRoot)\Temp\OfficeSetup"){
    Remove-Item -Path "$($env:SystemRoot)\Temp\OfficeSetup" -Recurse -Force -ErrorAction SilentlyContinue
}
Write-LogEntry -Value "Proofing Tools $($LanguageID) $($Action) completed" -Severity 1
